precompiles/ibc/ibc.go

```diff
@@ -0,0 +1,170 @@
+package ibc
+
+import (
+	"bytes"
+	"embed"
+	"errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	clienttypes "github.com/cosmos/ibc-go/v3/modules/core/02-client/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/vm"
+	pcommon "github.com/sei-protocol/sei-chain/precompiles/common"
+	"math/big"
+)
+
+const (
+	TransferMethod = "transfer"
+)
+
+const (
+	IBCAddress = "0x0000000000000000000000000000000000001008"
+)
+
+var _ vm.PrecompiledContract = &Precompile{}
+
+// Embed abi json file to the executable binary. Needed when importing as dependency.
+//
+//go:embed abi.json
+var f embed.FS
+
+func GetABI() abi.ABI {
+	abiBz, err := f.ReadFile("abi.json")
+	if err != nil {
+		panic(err)
+	}
+
+	newAbi, err := abi.JSON(bytes.NewReader(abiBz))
+	if err != nil {
+		panic(err)
+	}
+	return newAbi
+}
+
+type Precompile struct {
+	pcommon.Precompile
+	address        common.Address
+	transferKeeper pcommon.TransferKeeper
+	evmKeeper      pcommon.EVMKeeper
+
+	TransferID []byte
+}
+
+func NewPrecompile(transferKeeper pcommon.TransferKeeper, evmKeeper pcommon.EVMKeeper) (*Precompile, error) {
+	newAbi := GetABI()
+
+	p := &Precompile{
+		Precompile:     pcommon.Precompile{ABI: newAbi},
+		address:        common.HexToAddress(IBCAddress),
+		transferKeeper: transferKeeper,
+		evmKeeper:      evmKeeper,
+	}
+
+	for name, m := range newAbi.Methods {
+		switch name {
+		case TransferMethod:
+			p.TransferID = m.ID
+		}
+	}
+
+	return p, nil
+}
+
+// RequiredGas returns the required bare minimum gas to execute the precompile.
+func (p Precompile) RequiredGas(input []byte) uint64 {
+	methodID := input[:4]
+
+	method, err := p.ABI.MethodById(methodID)
+	if err != nil {
+		// This should never happen since this method is going to fail during Run
+		return 0
+	}
+
+	return p.Precompile.RequiredGas(input, p.IsTransaction(method.Name))
+}
+
+func (p Precompile) Run(evm *vm.EVM, caller common.Address, input []byte, value *big.Int) (bz []byte, err error) {
+	ctx, method, args, err := p.Prepare(evm, input)
+	if err != nil {
+		return nil, err
+	}
+
+	switch method.Name {
+	case TransferMethod:
+		return p.transfer(ctx, method, args)
+	}
+	return
+}
+func (p Precompile) transfer(ctx sdk.Context, method *abi.Method, args []interface{}) ([]byte, error) {
+	pcommon.AssertArgsLength(args, 6)
```

Comment:
@Kbhat1 @codchen do we need to do other validations?

Reply:
I see that `from` is set an arg. Can anyone trigger an IBC transfer on anyone else's behalf? If not then we should  be looking at  `caller` in https://github.com/sei-protocol/sei-chain/pull/1459/files#diff-3c39140061c990bd99f477fe4dbc73c0eaac283bfa477c1ba439016daf6fe688R87 to use as the from address instead

Reply:
Can reference https://github.com/sei-protocol/sei-chain/blob/seiv2/precompiles/bank/bank.go#L116 for example usage of `caller`